### PR TITLE
group records by keys of record

### PR DESCRIPTION
**Pull request by woutersm**

### CODE REVIEW

File: target_snowflake/__init__.py [hunk 1]
```diff
@@ -431,13 +431,64 @@ def load_stream_batch(stream, records, row_count, db_sync, no_compression=False,
...
+
 
 def flush_records(stream: str,
                   records: List[Dict],
                   db_sync: DbSync,
                   temp_dir: str = None,
                   no_compression: bool = False,
-                  archive_load_files: Dict = None) -> None:
+                  archive_load_files: Dict = None,
+                  schema = None) -> None:
```
Comment (mvgijssel) [inline target_snowflake/__init__.py]: What happens if the schema is `None`?

File: target_snowflake/__init__.py [hunk 2]
```diff
@@ -431,13 +431,64 @@ def load_stream_batch(stream, records, row_count, db_sync, no_compression=False,
...
+        del schema[unwanted_key]
+
+    return schema
+
+
+# group records by unique combination of keys of a record
+def group_records(records, full_schema):
+    grouped_records = {}
+    
+    for record_key, record in records.items():
```
Comment (mvgijssel) [inline target_snowflake/__init__.py]: Curious what the the `record_key`. Is that the primary key?

File: target_snowflake/__init__.py [hunk 3]
```diff
@@ -431,13 +431,64 @@ def load_stream_batch(stream, records, row_count, db_sync, no_compression=False,
...
+      key = "-".join((list(record.keys())))
+      schema = schema_from_record(record, full_schema)
+
+      if key in grouped_records:
+        grouped_records[key]["records"][record_key] = record
+      else:
+        grouped_record[key] = {"records": {record_key: record}, "schema": schema})
```
Comment (mvgijssel) [inline target_snowflake/__init__.py]: Don't know if Python supports it, but you can put the record keys in a `Set` object and use the set as the key of the grouped_records dict. That way you don't have to worry about, for example, the order of the keys of the record!

File: target_snowflake/__init__.py [hunk 4]
```diff
@@ -431,13 +431,64 @@ def load_stream_batch(stream, records, row_count, db_sync, no_compression=False,
         # reset row count for the current stream
         row_count[stream] = 0
 
+# create schema matching record (deleting unwanted keys)
+def schema_from_record(record, schema):
+    unwanted = set(schema) - set(record)
```
Comment (mvgijssel) [inline target_snowflake/__init__.py]: What are the shapes of the schema and the record here?

File: target_snowflake/__init__.py (hunk 1 above)
Comment (woutersm) [reply target_snowflake/__init__.py]: Oh missed that, `db_sync.flatted_schema` should be used, will update.

File: target_snowflake/__init__.py (hunk 3 above)
Comment (woutersm) [reply target_snowflake/__init__.py]: A non-hashable object can not be a key. But I can  try `frozenset`.

File: target_snowflake/__init__.py (hunk 4 above)
Comment (woutersm) [reply target_snowflake/__init__.py]: Set modifies the dicts to an interator of the keys of the dict, so the shapes don't really matter. I'll make it more explicit by calling .keys() on the objects first.

File: target_snowflake/__init__.py (hunk 2 above)
Comment (woutersm) [reply target_snowflake/__init__.py]: Yes, it is.

**Review by mvgijssel — APPROVED** [review]

**Pull request merged**